annotation-utils/src/main/java/edu/isi/vista/annotationutils/ExtractAnnotationStats.kt

```diff
@@ -110,22 +136,150 @@ class ExtractAnnotationStats {
+                                            )
+                                        }
+                                        documentAnnotatedSentences.add(sentenceID)
+                                        // We've found the sentence of our span
+                                        break
+                                    }
+                                }
                             }
                         }
                     }
-            return docsList
+            return sentenceList
+        }
+
+        /**
+         * Determine which spans in an ACE document are clue words
+         * or secondary triggers - these have been added by an annotator.
+         */
+        private fun getAceAdditions(relations: JsonNode, spans: JsonNode): List<JsonNode> {
+            val addedSpans: MutableSet<JsonNode> = mutableSetOf()
+            for (relation in relations) {
+                val relationType = if (relation["relation_type"] == null) {
+                    // Hack for some older ACE documents where
+                    // no label was added to clue words
+                    INTERESTING_ROLE
+                } else {
+                    relation["relation_type"].toString()
+                }
+                if (relationType == INTERESTING_ROLE || relationType == TRIGGER_ROLE) {
+                    addedSpans.add(spans[relation["Governor"].toString()])
+                }
+            }
+            return addedSpans.toList()
+        }
+
+        /**
+         * Determine which spans are trigger spans
+         *
+         * Input should not include singleton spans
+         * (i.e. the input spans here should all be part of relations).
+         * Spans not in relations are collected by another method.
+         */
+        private fun getPrimarySpans(relations: JsonNode, spans: JsonNode): List<JsonNode> {
+            val dependents: MutableSet<String> = mutableSetOf()
+            val governors: MutableSet<String> = mutableSetOf()
+            val primarySpans: MutableSet<JsonNode> = mutableSetOf()
+            for (relation in relations) {
+                val relationDependent = relation["Dependent"].toString()
+                val relationGovernor = relation["Governor"].toString()
+                // A span is a primary trigger if it
+                // depends on nothing or itself.
+                // We prevent spans from being added to
+                // the "governors" list to ensure that
+                // triggers which serve as their own
+                // arguments are still considered primary triggers.
+                dependents.add(relationDependent)
+                if (relationDependent != relationGovernor) {
+                    governors.add(relationGovernor)
+                }
+            }
+            for (dependent in dependents) {
+                if (!governors.contains(dependent)) {
+                    primarySpans.add(spans[dependent])
+                }
```

Comment:
Adding a dependent to `primarySpans` if it's not in the set of governors in the first loop wouldn't necessarily filter out all secondary triggers.

For example, if there are two spans A and B where A is an argument of itself and a secondary trigger to B, combining these loops would incorrectly make A a primary trigger (Two relations: (dep=A, gov=A), (dep=B, gov=A)).

However, it *should* still yield the same sentence count since there is a check for duplicate sentences - it would just have another trigger to process.

Reply:
This was hard for me to understand without writing out explicit pseudocode. Does @gabbard have an opinion on if there should be more of a comment here?

Reply:
I think this is better, and if other people want to see it, they can write out all the code with explicit values like I did.